src/libraries/System.Linq/src/System/Linq/Sequence.cs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        /// <summary>Generates a  sequence that begins with <paramref name="start"/> and yields additional values each incremented by <paramref name="step"/> until <paramref name="endInclusive"/> is reached.</summary>
+        /// <typeparam name="T">The type of the value to be yielded in the result sequence.</typeparam>
+        /// <param name="start">The starting value. This value will always be included in the resulting sequence.</param>
+        /// <param name="endInclusive">The ending bound beyond which values will not be included in the sequence.</param>
+        /// <param name="step">The amount by which the next value in the sequence should be incremented from the previous value.</param>
+        /// <returns>An <see cref="IEnumerable{T}"/> that contains the sequence.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="start"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="step"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="endInclusive"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="step"/> is greater than zero but <paramref name="endInclusive"/> is less than <paramref name="start"/>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="step"/> is less than zero but <paramref name="endInclusive"/> is greater than <paramref name="start"/>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="step"/> is zero and <paramref name="endInclusive"/> does not equal <paramref name="start"/>.</exception>
+        public static IEnumerable<T> Sequence<T>(T start, T endInclusive, T step) where T : INumber<T>
+        {
+            if (start is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.start);
+            }
+
+            if (endInclusive is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.endInclusive);
+            }
+
+            if (step is null)
```

Comment:
Should this be detecting and accounting for `NaN`?

We likely don't need to handle `Infinite` since that is comparable and will modify the value.

Reply:
> Should this be detecting and accounting for NaN?

What should it do for NaN... throw?